tools/release.go

```diff
@@ -0,0 +1,154 @@
+//   named release tag and a Patch that is one greater, and push that tag.
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/mod/semver"
+)
+
+func runOrDie(cmd *exec.Cmd) string {
+	fmt.Println(cmd.String())
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Print(string(out))
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	return string(out)
+}
+
+func main() {
+	if len(os.Args) >= 2 && os.Args[1] == "hotfix" {
+		os.Args = append([]string{os.Args[0]}, os.Args[2:]...)
+		hotfix()
+	} else {
+		release()
+	}
+}
+
+func release() {
+	var push bool
+	flag.BoolVar(&push, "f", false, "If set, push the resulting release tag to GitHub.")
+	flag.Parse()
+
+	// Fetch all of the latest commits and refs from origin, so that we can ensure
+	// we're tagging the correct commit and not recreating an existing tag.
+	runOrDie(exec.Command("git", "fetch", "origin"))
+
+	// We use semver's vMajor.Minor.Patch format, where the Major version is
+	// always 0 (no backwards compatibility guarantees), the Minor version is
+	// the date of the release, and the Patch number is zero for normal releases
+	// and only non-zero for hotfix releases.
+	minor := time.Now().Format("20060102")
+	version := fmt.Sprintf("v0.%s.0", minor)
+	message := fmt.Sprintf("Release %s", version)
+
+	// Produce the tag, using -s to PGP sign it. This will fail if a tag with
+	// that name already exists.
+	runOrDie(exec.Command("git", "tag", "-s", "-m", message, version, "origin/main"))
+
+	// Show the result of the tagging operation, including the tag message and
+	// signature, and the commit hash and message, but not the diff.
+	show := runOrDie(exec.Command("git", "show", "-s", version))
+	fmt.Print(show)
+
+	if push {
+		runOrDie(exec.Command("git", "push", "origin", version))
+	} else {
+		fmt.Printf("created tag %s, exiting\n", version)
+	}
+}
+
+func hotfix() {
+	var cherryPicks string
+	flag.StringVar(&cherryPicks, "c", "", "Comma-separated list of commit hashes to cherry-pick on top of the prior release.")
+	var onto string
+	flag.StringVar(&onto, "o", "", "Name of the existing release tag to base this hotfix on top of.")
+	var push bool
+	flag.BoolVar(&push, "f", false, "If set, push the resulting release tag to GitHub.")
+	flag.Parse()
+
+	if !semver.IsValid(onto) {
+		fmt.Printf("Tag name %q is not a valid release tag\n", onto)
+		os.Exit(1)
+	}
+
+	if len(cherryPicks) == 0 {
+		fmt.Printf("You must specify one or more commits to cherry-pick\n")
+		os.Exit(1)
+	}
+
+	// Fetch all of the latest commits and refs from origin, so that we can ensure
+	// we're hotfixing an existing release and cherry-picking on to the correct
+	// branch.
+	runOrDie(exec.Command("git", "fetch", "origin"))
+
+	// Confirm that the release tag we're cherry-picking onto actually exists and
+	// is a tag.
+	runOrDie(exec.Command("git", "rev-parse", "--verify", fmt.Sprintf("%s^{tag}", onto)))
```

Comment:
Fixes gosec `G204 Subprocess launched with a potential tainted input or cmd arguments`.
```suggestion
	arg := fmt.Sprintf("%s^{tag}", onto)
	runOrDie(exec.Command("git", "rev-parse", "--verify", arg))
```

Reply:
I really dislike that this works. It's exploiting a weakness in the linter, rather than actually fixing the underlying issue. I'm trying to see if there's a better way.

Reply:
From conversation outside of github, this is going to be switched to ignoring the G204 lint.

Reply:
I changed my mind back, because it's not easy to only ignore a single gosec lint, you can only `//nolint:gosec` and ignore _all_ gosec lints.